### PR TITLE
rPackages.rsbml: fix build

### PR DESCRIPTION
**Pull request by Kupac**
## Description of changes

Packaged libSBML and added it as a buildInput for rsbml

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Kupac** [comment]
@ofborg build rPackages.rsbml

**Comment by Kupac** [comment]
Could someone check what's wrong with libsbml on darwin? @nviets  @philipp-baumann

**Comment by philipp-baumann** [comment]
on `aarch64-darwin`, I get

```r
➜ nix-shell -p R rPackages.rsbml               
this derivation will be built:
  /nix/store/dzcyav42gghmll5r0wmq41xf7n005vd6-r-rsbml-2.60.0.drv
building '/nix/store/dzcyav42gghmll5r0wmq41xf7n005vd6-r-rsbml-2.60.0.drv'...
Running phase: unpackPhase
unpacking source archive /nix/store/dbk270nm37zyzhvp35vp6n3746ilvzsx-rsbml_2.60.0.tar.gz
source root is rsbml
setting SOURCE_DATE_EPOCH to timestamp 1698187740 of file rsbml/inst/SBML_odeSolver-1.7.0beta/src/sbmlsolver/ASTCompiler/Makefile
Running phase: patchPhase
Running phase: updateAutotoolsGnuConfigScriptsPhase
Updating Autotools / GNU config script to a newer upstream version: ./inst/SBML_odeSolver-1.7.0beta/config/config.sub
Updating Autotools / GNU config script to a newer upstream version: ./inst/SBML_odeSolver-1.7.0beta/config/config.guess
Running phase: configurePhase
Running phase: buildPhase
Running phase: checkPhase
Running phase: installPhase
During startup - Warning messages:
1: Setting LC_CTYPE failed, using "C" 
2: Setting LC_TIME failed, using "C" 
3: Setting LC_MESSAGES failed, using "C" 
4: Setting LC_MONETARY failed, using "C" 
* installing *source* package 'rsbml' ...
** using staged installation
checking for pkg-config... no
checking for LIBSBML... no
configure: error: in `/private/tmp/nix-build-r-rsbml-2.60.0.drv-0/rsbml':
configure: error: The pkg-config script could not be found or is too old.  Make sure it
is in your PATH or set the PKG_CONFIG environment variable to the full
path to pkg-config.

Alternatively, you may set the environment variables LIBSBML_CFLAGS
and LIBSBML_LIBS to avoid the need to call pkg-config.
See the pkg-config man page for more details.

To get pkg-config, see <http://pkg-config.freedesktop.org/>.
See `config.log' for more details
ERROR: configuration failed for package 'rsbml'
* removing '/nix/store/8fxzn1skq4kklqisrfj8c4kb20skwhkr-r-rsbml-2.60.0/library/rsbml'
error: builder for '/nix/store/dzcyav42gghmll5r0wmq41xf7n005vd6-r-rsbml-2.60.0.drv' failed with exit code 1;
       last 10 log lines:
       > path to pkg-config.
       >
       > Alternatively, you may set the environment variables LIBSBML_CFLAGS
       > and LIBSBML_LIBS to avoid the need to call pkg-config.
       > See the pkg-config man page for more details.
       >
       > To get pkg-config, see <http://pkg-config.freedesktop.org/>.
       > See `config.log' for more details
       > ERROR: configuration failed for package 'rsbml'
       > * removing '/nix/store/8fxzn1skq4kklqisrfj8c4kb20skwhkr-r-rsbml-2.60.0/library/rsbml'
       For full logs, run 'nix log /nix/store/dzcyav42gghmll5r0wmq41xf7n005vd6-r-rsbml-2.60.0.drv'.
```

**Comment by philipp-baumann** [comment]
> Could someone check what's wrong with libsbml on darwin? @nviets @philipp-baumann

for `libsbml`. does it need a `callPackage` approach?

```sh
nixpkgs on  fix_rsbml [$?] 
➜ nix-shell -p libsbml     
error:
       … while calling the 'derivationStrict' builtin

         at /builtin/derivation.nix:9:12: (source not available)

       … while evaluating derivation 'shell'
         whose name attribute is located at /nix/store/6nnv8z57j976ij3img6ni46fkbac9vb0-nixpkgs/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:331:7

       … while evaluating attribute '__impureHostDeps' of derivation 'shell'

         at /nix/store/6nnv8z57j976ij3img6ni46fkbac9vb0-nixpkgs/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:443:7:

          442|       __propagatedSandboxProfile = unique (computedPropagatedSandboxProfile ++ [ propagatedSandboxProfile ]);
          443|       __impureHostDeps = computedImpureHostDeps ++ computedPropagatedImpureHostDeps ++ __propagatedImpureHostDeps ++ __impureHostDeps ++ stdenv.__extraImpureHostDeps ++ [
             |       ^
          444|         "/dev/zero"

       error: undefined variable 'libsbml'

       at «string»:1:107:

            1| {...}@args: with import <nixpkgs> args; (pkgs.runCommandCC or pkgs.runCommand) "shell" { buildInputs = [ (libsbml) ]; } ""
             | 
```

**Comment by philipp-baumann** [comment]
For `libsbml`, i get this on `aarch64-darwin`:

```sh
-- The C compiler identification is Clang 16.0.6
-- The CXX compiler identification is Clang 16.0.6
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /nix/store/v4p0piwwk3a9gwrhvxa8b04ljd1arviz-clang-wrapper-16.0.6/bin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /nix/store/v4p0piwwk3a9gwrhvxa8b04ljd1arviz-clang-wrapper-16.0.6/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working C compiler: /nix/store/v4p0piwwk3a9gwrhvxa8b04ljd1arviz-clang-wrapper-16.0.6/bin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Found SWIG: /nix/store/ada41dic6af4bkkpa1f7178jm7a418bh-swig-3.0.12/bin/swig (found version "3.0.12")
-- Found PkgConfig: /nix/store/3l8pzzkvvihqj5z1bymn5zcsm1a3gmp9-pkg-config-wrapper-0.29.2/bin/pkg-config (found version "0.29.2")
-- Looking for sin in m
-- Looking for sin in m - found
-- Performing Test LIBXML_LIBXML_TEST
-- Performing Test LIBXML_LIBXML_TEST - Failed
-- LIBXML_LIBXML_TEST = 
-- Performing Test LIBXML_LIBXML_TEST2
-- Performing Test LIBXML_LIBXML_TEST2 - Failed
-- LIBXML_LIBXML_TEST2 = 
CMake Error at CMakeModules/FindLIBXML.cmake:142 (message):
  Unable to compile a test executable against LIBXML



      LIBXML_INCLUDE_DIR = /nix/store/1fzfrlh08afwsc5dr40bmr1cra1y6kx6-libxml2-2.12.6-dev/include/libxml2
      LIBXML_LIBRARY     = /nix/store/69p1zrdp1648ls30sn21v99v8bwapv5c-libxml2-2.12.6/lib/libxml2.so

  

      
Call Stack (most recent call first):
  CMakeLists.txt:446 (find_package)


-- Configuring incomplete, errors occurred!
error: builder for '/nix/store/wm00m4scqnlk8yvjs62z3jf9dg027g0b-libsbml-5.20.2.drv' failed with exit code 1;
       last 10 log lines:
       >       LIBXML_LIBRARY     = /nix/store/69p1zrdp1648ls30sn21v99v8bwapv5c-libxml2-2.12.6/lib/libxml2.so
       >
       >
       >
       >
       > Call Stack (most recent call first):
       >   CMakeLists.txt:446 (find_package)
       >
       > 
       > -- Configuring incomplete, errors occurred!
       For full logs, run 'nix log /nix/store/wm00m4scqnlk8yvjs62z3jf9dg027g0b-libsbml-5.20.2.drv'.
```

**Comment by Kupac** [comment]
Thanks for testing, I saw the same error from ofborg, but the logs don't tell me what's missing on darwin. The xml tests run fine on linux.  Any clue?

**Review by b-rodrigues — APPROVED** [review]
can confirm it builds and runs on aarch64-linux

LGTM

**Comment by b-rodrigues** [comment]
> Thanks for testing, I saw the same error from ofborg, but the logs don't tell me what's missing on darwin. The xml tests run fine on linux. Any clue?

They say the darwin binaries are  _built with the Apple-supplied version of the [http://www.xmlsoft.org](https://sourceforge.net/projects/sbml/files/libsbml/5.19.0/stable/Mac%20OS%20X/libxml2) parser library,_
see: https://sourceforge.net/projects/sbml/files/libsbml/5.19.0/stable/Mac%20OS%20X/

could this be it?

**Comment by Kupac** [comment]
Ah, good catch! But thats not the case on nix, we use thr same xml lib as in linux. Maybe there's something in the test code that expects the apple lib on darwin.

**Comment by philipp-baumann** [comment]
> Ah, good catch! But thats not the case on nix, we use thr same xml lib as in linux. Maybe there's something in the test code that expects the apple lib on darwin.

yes it seems in test code: 

> Ah, good catch! But thats not the case on nix, we use thr same xml lib as in linux. Maybe there's something in the test code that expects the apple lib on darwin.

seems to happen here according to the logs: https://github.com/sbmlteam/libsbml/blob/ec444e9eab9a94ca84d23efbc0394db69536beca/CMakeModules/FindLIBXML.cmake#L137C1-L142

**Comment by Kupac** [comment]
OK, let's try again with static xml lib on darwin. Fingers crossed.
@ofborg build rPackages.rsbml

**Comment by Kupac** [comment]
Meh, static xml didn't help the darwin build :-/ why do those tests fail on darwin, and not on linux?

**Comment by Kupac** [comment]
Guys, I don't think I can do much more for the darwin build, inless someone with hardware can help out. But at least it's a partial fix for linux. We can extend to darwin later. Can we go ahead and merge this? Should I change meta.platforms to linux only?

**Comment by philipp-baumann** [comment]
> Guys, I don't think I can do much more for the darwin build, inless someone with hardware can help out. But at least it's a partial fix for linux. We can extend to darwin later. Can we go ahead and merge this? Should I change meta.platforms to linux only?

Sounds ok to me. I currently don't have the time to try-error dig into patching cmake config for darwin build, to not block things further i think its ok to merge from my side and extend to darwin later.

**Comment by Kupac** [comment]
@ofborg build rPackages.rsbml

**Review by jbedo — DISMISSED** [review]

File: pkgs/by-name/li/libsbml/package.nix [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+  buildInputs = [
+    bzip2.dev
+    libxml2
+  ];
+
+  # libSBML doesn't always make use of pkg-config
+  cmakeFlags = [
+    "-DLIBXML_INCLUDE_DIR=${lib.getDev libxml2}/include/libxml2"
+    "-DLIBXML_LIBRARY=${lib.getLib libxml2}/lib/libxml2.so"
```
Comment (jbedo) [inline pkgs/by-name/li/libsbml/package.nix:46]:
```suggestion
    "-DLIBXML_LIBRARY=${lib.getLib libxml2}/lib/libxml2.${stdenv.hostPlatform.extensions.sharedLibrary}"
```
I think this should resolve the darwin issues.

**Comment by Kupac** [comment]
@ofborg build rPackages.rsbml

**Comment by Kupac** [comment]
@ofborg build rPackages.rsbml

**Comment by Kupac** [comment]
Thanks for the tip, @jbedo , that worked! After ofborg built it on x86_64-darwin, I rebased the changes, so now we have just the two commits. Cheers!

**Review by jbedo — APPROVED** [review]

**Pull request merged**